app/src/commonMain/kotlin/io/github/cgr/app/Program.kt

```diff
@@ -0,0 +1,17 @@
+package io.github.cgr.app
+
+import io.github.cgr.app.objects.Opcode
+
+data class Program(val opcodes: Array<Pair<Opcode, IntArray>>, val constantPool: Array<Any?>) {
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is Program) return false
+        return opcodes.contentEquals(other.opcodes) && constantPool.contentEquals(other.constantPool)
+    }
+
+    override fun hashCode(): Int {
+        var result = opcodes.hashCode()
+        result = 31 * result + constantPool.contentHashCode()
```

Comment:
I'm curious where the 31 comes from here.

Alternatively, does Kotlin have an equivalent to Swift's `Hasher`? In Swift you might write

```swift
var hasher = Hasher()
hasher.combine(opcodes)
hasher.combine(constantPool)
return hasher.finalize()
```

Reply:
yeahs `Objects.hash` exists but its generally not recommended because varargs. This `hashCode` was IDE implemented. As to why 31, see [this SO answer](https://stackoverflow.com/a/299748/18162944).